File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2024/Y24D09.kt
package org.gristle.pdxpuzzles.everybodycodes.y2024

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.getIntList
import kotlin.math.ceil

object Y24D09 : Day {
    override fun part1(input: String): Int {
        val stamps = listOf(1, 3, 5, 10).reversed()
        val brightnesses: List<Int> = input.getIntList()
        val brightest: Int = brightnesses.max()
        val cache: MutableList<Int?> = MutableList(brightest + 1) { null }
        return solve(brightnesses, stamps, cache)
    }

    override fun part2(input: String): Int {
        val stamps = listOf(1, 3, 5, 10, 15, 16, 20, 24, 25, 30).reversed()
        val brightnesses: List<Int> = input.getIntList()
        val brightest: Int = brightnesses.max()
        val cache: MutableList<Int?> = MutableList(brightest + 1) { null }
        return solve(brightnesses, stamps, cache)
    }

    override fun part3(input: String): Int {
        val stamps = listOf(1, 3, 5, 10, 15, 16, 20, 24, 25, 30, 37, 38, 49, 50, 74, 75, 100, 101).reversed()
        val brightnesses: List<Int> = input.getIntList()
        val brightest: Int = brightnesses.max()
        val cache: MutableList<Int?> = MutableList(brightest + 1) { null }
        for (stamp in stamps) {
            cache[stamp] = 1
        }
        return brightnesses.sumOf { brightness ->
            val half = brightness / 2
            val limit = if (brightness and 1 == 1) 49 else 50
            (0..limit)
                .minOf { n ->
                    val ba = half - n
                    val bb = (brightness - half) + n
                    val ba_beetles = cache[ba] ?: getBeetles(ba, Int.MAX_VALUE, cache, stamps)
                    val bb_beetles = cache[bb] ?: getBeetles(bb, Int.MAX_VALUE, cache, stamps)
                    ba_beetles + bb_beetles
                }
        }
    }

    fun getBeetles(
        remaining: Int,
        prevBest: Int,
        cache: MutableList<Int?>,
        stamps: List<Int>
    ): Int = cache[remaining] ?: run {
        var best = prevBest
        outer@for ((idx, stamp) in stamps.withIndex()) {
            val beetles = remaining / stamp
            if (beetles >= best) return@getBeetles best
            if (remaining % stamp == 0) {
                if (cache[remaining] == null) {
                    cache[remaining] = beetles
                }
                return@getBeetles beetles
            }

            for (n in beetles downTo 1) {
                val nextStamp = stamps.getOrNull(idx + 1)
                if (nextStamp != null) {
                    val nextStampBeetles = ceil((remaining - n * stamp).toFloat() / nextStamp.toFloat()).toInt()
                    if (n + nextStampBeetles >= best) break@outer
                }
                if (best <= n) { continue }
                val result = n + getBeetles(
                    remaining - n * stamp,
                    best - n,
                    cache,
                    stamps.drop(idx)
                )
                if (result < best) {
                    best = result
                }
            }
        }
        cache[remaining] = best
        best
    }


    private fun solve(brightnesses: List<Int>, stamps: List<Int>, cache: MutableList<Int?>): Int {
        for (stamp in stamps) {
            cache[stamp] = 1
        }
        return brightnesses
            .sumOf { brightness ->
                cache[brightness] ?: getBeetles(brightness, Int.MAX_VALUE, cache, stamps)
            }
    }

}

fun main() = Day.runDay(Y24D09::class)

//    Quest 1: 12442 (1ms)
//    Quest 2: 5071 (3ms)
//    Quest 3: 151150 (11ms)
//    Total time: 16ms

File: rust/everybody-codes/src/bin/2024/y24d09.rs
use everybody_codes::utilities::inputs::get_inputs;
use utilities::{
    parsing::{get_numbers::ContainsNumbers, try_get::TryGet},
    structs::stopwatch::{ReportDuration, Stopwatch},
};

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_inputs(24, 9);
    println!("Inputs loaded ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn part1(input: &str) -> usize {
    let stamps: Vec<usize> = [1, 3, 5, 10].into_iter().rev().collect();
    let brightnesses = get_brightnesses(input);
    let &brightest = brightnesses.iter().max().unwrap();
    let mut cache: Vec<Option<usize>> = vec![None; brightest + 1];
    solve(&brightnesses, &stamps, &mut cache)
}

fn part2(input: &str) -> usize {
    let stamps: Vec<usize> = [1, 3, 5, 10, 15, 16, 20, 24, 25, 30]
        .into_iter()
        .rev()
        .collect();
    let brightnesses = get_brightnesses(input);
    let &brightest = brightnesses.iter().max().unwrap();
    let mut cache: Vec<Option<usize>> = vec![None; brightest + 1];
    solve(&brightnesses, &stamps, &mut cache)
}

fn part3(input: &str) -> usize {
    let stamps: Vec<usize> = [
        1, 3, 5, 10, 15, 16, 20, 24, 25, 30, 37, 38, 49, 50, 74, 75, 100, 101,
    ]
    .into_iter()
    .rev()
    .collect();
    let brightnesses = get_brightnesses(input);
    let &brightest = brightnesses.iter().max().unwrap();
    let mut cache: Vec<Option<usize>> = vec![None; brightest + 1];
    for stamp in stamps.iter() {
        cache[*stamp] = Some(1);
    }
    brightnesses
        .iter()
        .map(|brightness| {
            let half = brightness / 2;
            let limit = if brightness & 1 == 1 { 49 } else { 50 };
            (0..=limit)
                .map(|n| {
                    let ba = half - n;
                    let bb = (brightness - half) + n;
                    let ba_beetles = if let Some(ba_beetles) = cache[ba] {
                        ba_beetles
                    } else {
                        get_beetles(ba, usize::MAX, &mut cache, &stamps)
                    };
                    let bb_beetles = if let Some(bb_beetles) = cache[bb] {
                        bb_beetles
                    } else {
                        get_beetles(bb, usize::MAX, &mut cache, &stamps)
                    };
                    ba_beetles + bb_beetles
                })
                .min()
                .unwrap()
        })
        .sum()
}

fn get_brightnesses(input: &str) -> Vec<usize> {
    input.get_numbers().collect()
}

fn solve(brightnesses: &[usize], stamps: &[usize], cache: &mut Vec<Option<usize>>) -> usize {
    for stamp in stamps {
        cache[*stamp] = Some(1);
    }
    brightnesses
        .iter()
        .map(|brightness| {
            if let Some(beetles) = cache[*brightness] {
                beetles
            } else {
                get_beetles(*brightness, usize::MAX, cache, stamps)
            }
        })
        .sum()
}

fn get_beetles(
    remaining: usize,
    prev_best: usize,
    cache: &mut Vec<Option<usize>>,
    stamps: &[usize],
) -> usize {
    if let Some(count) = cache[remaining] {
        return count;
    }

    // otherwise, carry on
    let mut best = prev_best;

    // iterate through each stamp in descending order
    'outer: for (idx, &stamp) in stamps.iter().enumerate() {
        let beetles = remaining / stamp;
        // if the best possible outcome requires more than the current best, abort
        if beetles >= best {
            return best;
        }

        // if divides cleanly, we have our answer, update cache and pop up with it
        if remaining % stamp == 0 {
            // may not need first conditional
            if cache[remaining].is_none() {
                cache[remaining] = Some(beetles);
            }
            return beetles;
        }

        // otherwise try decreasing amounts of that stamp
        // only runs if the stamp is smaller than the remaining brightness
        for n in (1..=beetles).rev() {
            // there should be break here where we know not to continue because anything
            // remaining will result in something bigger than current best.
            // maybe get the next stamp and do the calculation there...
            if let Some(&next_stamp) = stamps.try_get(idx + 1) {
                let next_stamp_beetles =
                    ((remaining - n * stamp) as f32 / next_stamp as f32).ceil() as usize;
                if n + next_stamp_beetles >= best {
                    break 'outer;
                }
            }
            // this should be the best obtainable with n number of stamp
            if best <= n {
                continue;
            }
            let result = n + get_beetles(remaining - n * stamp, best - n, cache, &stamps[idx..]);
            if result < best {
                best = result;
            }
        }
    }
    cache[remaining] = Some(best);
    best
}

// #[test]
// fn tests() {
//     let tests = [r"2, 4, 7, 16", "27", "156488, 352486, 546212"];
//     let stamps = [vec![1, 3, 5, 10], vec![1, 3, 9, 10], vec![1, 3, 5, 10, 15, 16, 20, 24, 25, 30, 37, 38, 49, 50, 74, 75, 100, 101]];
//     assert_eq!(10, part1(tests[0], &stamps[0]));
//     assert_eq!(3, solve(tests[1], &stamps[1]));
//     assert_eq!(10449, part3(tests[2], &stamps[2]));
// }
